View.elm

```diff
@@ -0,0 +1,304 @@
+module View exposing (..)
+
+import Html exposing (Html)
+import Html.Events
+
+
+-- # Elements
+-- This is a type level API for Html documents.
+-- Use the standard types defined here or create your own.
+
+
+type H1 child
+    = H1Type Never
+
+
+type H2 child
+    = H2Type Never
+
+
+type Section child
+    = SectionType Never
+
+
+type P child
+    = PType Never
+
+
+type OnClick msg child
+    = OnClickType Never
+
+
+type OneOf child1 child2
+    = OneOfType Never
+
+
+
+-- # Single type representation
+-- This is meant for creating view functions.
+-- 1. A phantom type ensures the View corresponds to the page type.
+-- 2. Customize the `SubView` type with your own stuff.
+-- 3. We don't render directly into Html to support different interpreters:
+--    - Elm-css uses a different Html type.
+--    - A view test might wish to interpret the view in a different way.
+--    - Certain optimizations (like collapsing unnecessary containers) are only
+--      possible if we can introspect our page structure.
+
+
+type View tipe custom msg
+    = View (SubView custom msg)
+
+
+type SubView custom msg
+    = H1 (SubView custom msg)
+    | H2 (SubView custom msg)
+    | Section (SubView custom msg)
+    | P (SubView custom msg)
+    | Text String
+    | List (List (SubView custom msg))
+    | OnClick msg (SubView custom msg)
```

Comment:
💭 maybe a separate type for attributes.

Reply:
That makes sense, we should definitely support arbitrary attributes. I'll have to think a bit more about the best way to do it though. The closer we move this type towards `Html` the less work it is to support all uses of Html, but also the less information we have in the `View` type that can be introspected. For instance: If we'd build a 'test' interpreter on top of this structure that would need to handle clicks, it would be able to recognize an explicit `OnClick` constructor, but not an `onClick` attribute hidden in a generic attribute constructor. I'd like to revisit this in a future iteration!